code/controllers/subsystems/machines.dm

```diff
@@ -110,67 +118,111 @@ datum/controller/subsystem/machines/proc/setup_atmos_machinery(list/machines)
 
 /datum/controller/subsystem/machines/proc/process_pipenets(resumed = 0)
 	if (!resumed)
-		src.current_run = pipenets.Copy()
-	//cache for sanic speed (lists are references anyways)
-	var/list/current_run = src.current_run
-	while(current_run.len)
-		var/datum/pipe_network/PN = current_run[current_run.len]
-		current_run.len--
-		if(istype(PN) && !QDELETED(PN))
-			PN.Process(wait)
+		nextProcessingListPosition = 1 //fresh start, otherwise from saved posisition
+
+	//localizations
+	var/list/local_list = pipenets
+	var/datum/pipe_network/thing
+	var/wait = src.wait
+
+	var/tickCheckPeriod = round(local_list.len/16+1) //pause process every 1/16th length of list
+	while(nextProcessingListPosition && (nextProcessingListPosition <= local_list.len)) //until position is valid
+		thing = local_list[nextProcessingListPosition]
+		nextProcessingListPosition++
+
+		if(!QDELETED(thing))
+			thing.Process(wait)
 		else
-			pipenets.Remove(PN)
-			PN.is_processing = null
-		if(MC_TICK_CHECK)
+			local_list.Remove(thing)
+			thing.is_processing = null
+			nextProcessingListPosition-- //removing processed thing from list moves the queue to the left, adjust accordingly
+
+		if(!(nextProcessingListPosition%tickCheckPeriod)) //pauses every tickCheckPeriod-th processed thing
+			pause()
 			return
 
+	nextProcessingListPosition = 0 //entire list was processed
+
 /datum/controller/subsystem/machines/proc/process_machinery(resumed = 0)
 	if (!resumed)
-		src.current_run = machinery.Copy()
-
-	var/list/current_run = src.current_run
-	while(current_run.len)
-		var/obj/machinery/M = current_run[current_run.len]
-		current_run.len--
-		if(istype(M) && !QDELETED(M) && !(M.Process(wait) == PROCESS_KILL))
-			if(M.use_power)
-				M.auto_use_power()
+		nextProcessingListPosition = 1 //fresh start, otherwise from saved posisition
+
+	//localizations
+	var/list/local_list = machinery
+	var/obj/machinery/thing
+	var/wait = src.wait
+
+	var/tickCheckPeriod = round(local_list.len/16+1) //pause process every 1/16th length of list
```

Comment:
Recalculating this every time you go to process doesn't seem right. What if the list has changed since the last time it was calculated

Reply:
The tick period can be an arbitrary number and it would not produce anomalous behavior. This is merely a guide for how many times it's supposed to pause based on current length of list.

Reply:
What about those 4 flags https://github.com/discordia-space/CEV-Eris/blob/master/code/__defines/MC.dm#L56 , is it necessary to add more methods to "pause" than adjusting flags?

Reply:
Idea was not to have it run a check on 8000 machines, and instead have it run only at most 16 times. CHECK_TICK might be just a relatively simple condition check, but it adds up into considerable overhead if it has to be done 8000 times.

Reply:
Okay.
Btw, we process every machines mostly not to process them, but for use_power proc to calculate power consumption on the maps (have you checked if there is any issues introduced with those pauses?).
If someone rework power consumtion calculations to be added and removed only when there is power change or dis/connections of machinery while normally we only work with cached value, then we don't need to add every machine into process list by default (it will become something similar as objects, when we only add things into process, where we actually need to process something), which should cut off needless processing by a lot i guess.
Also someone finally started to rework this on https://github.com/Baystation12/Baystation12/pull/23643 for areas, then it will be nice to do so for cables i think later.

Reply:
Reworking how power usage works is a good idea but not in scope of what I am doing in this PR, which is removing overhead from the subsystem itself rather than changing when and how things are processed.

Ultimately machines should be split into two categories, ones that are 'dumb' machines and only drain power with no other processing, and machines that actually require a process call periodically. Which is again outside of scope for this.